cli/src/commands/diff.rs

```diff
@@ -0,0 +1,526 @@
+        HashMap::new();
+    for constant in pallet_metadata_1.constants() {
+        let (e1, _) = constants.entry(constant.name()).or_default();
+        *e1 = Some(constant);
+    }
+    for constant in pallet_metadata_2.constants() {
+        let (e1, e2) = constants.entry(constant.name()).or_default();
+        // skip all entries with the same hash:
+        if let Some(e1_inner) = e1 {
+            let e1_hash = pallet_metadata_1
+                .constant_hash(e1_inner.name())
+                .expect("constant should be present");
+            let e2_hash = pallet_metadata_2
+                .constant_hash(constant.name())
+                .expect("constant should be present");
+            if e1_hash == e2_hash {
+                constants.remove(constant.name());
+                continue;
+            }
+        }
+        *e2 = Some(constant);
+    }
+    constants
+        .into_values()
+        .map(|tuple| Diff::try_from(tuple).unwrap())
+        .collect()
+}
+
+fn runtime_api_differences<'a>(
+    metadata_1: &'a Metadata,
+    metadata_2: &'a Metadata,
+) -> Vec<Diff<RuntimeApiMetadata<'a>>> {
+    let mut runtime_apis: HashMap<
+        &str,
+        (
+            Option<RuntimeApiMetadata<'a>>,
+            Option<RuntimeApiMetadata<'a>>,
+        ),
+    > = HashMap::new();
+
+    for runtime_api in metadata_1.runtime_api_traits() {
+        let (e1, _) = runtime_apis.entry(runtime_api.name()).or_default();
+        *e1 = Some(runtime_api);
+    }
+
+    for runtime_api in metadata_2.runtime_api_traits() {
+        let (e1, e2) = runtime_apis.entry(runtime_api.name()).or_default();
+        // skip all entries with the same hash:
+        if let Some(e1_inner) = e1 {
+            if e1_inner.hash() == runtime_api.hash() {
+                runtime_apis.remove(runtime_api.name());
+                continue;
+            }
+        }
+        *e2 = Some(runtime_api);
+    }
+    runtime_apis
+        .into_values()
+        .map(|tuple| Diff::try_from(tuple).unwrap())
+        .collect()
+}
+
+fn pallet_differences<'a>(
+    metadata_1: &'a Metadata,
+    metadata_2: &'a Metadata,
+) -> Vec<Diff<PalletMetadata<'a>>> {
+    let mut pallets: HashMap<&str, (Option<PalletMetadata<'a>>, Option<PalletMetadata<'a>>)> =
+        HashMap::new();
+
+    for pallet_metadata in metadata_1.pallets() {
+        let (e1, _) = pallets.entry(pallet_metadata.name()).or_default();
+        *e1 = Some(pallet_metadata);
+    }
+
+    for pallet_metadata in metadata_2.pallets() {
+        let (e1, e2) = pallets.entry(pallet_metadata.name()).or_default();
+        // skip all entries with the same hash:
+        if let Some(e1_inner) = e1 {
+            if e1_inner.hash() == pallet_metadata.hash() {
+                pallets.remove(pallet_metadata.name());
+                continue;
+            }
+        }
+        *e2 = Some(pallet_metadata);
+    }
+
+    pallets
+        .into_values()
+        .map(|tuple| Diff::try_from(tuple).unwrap())
```

Comment:
`Err()` should be unreachable based on the code above but would be good with an `expect` and a small explanation.